redash/tasks/queries.py

```diff
@@ -270,59 +272,62 @@ def get_table_sample_data(existing_columns, data_source_id, table_name, table_id
     )
     models.db.session.commit()
 
+
 def cleanup_data_in_table(table_model):
     removed_metadata = table_model.query.filter(
-        table_model.exists == False,
+        table_model.exists.is_(False),
     ).options(load_only('updated_at'))
 
     for removed_metadata_row in removed_metadata:
         is_old_data = (
             utils.utcnow() - removed_metadata_row.updated_at
         ) > datetime.timedelta(days=settings.SCHEMA_METADATA_TTL_DAYS)
 
-        table_model.query.filter(
-            table_model.id == removed_metadata_row.id,
-        ).delete()
+        if is_old_data:
+            table_model.query.filter(
+                table_model.id == removed_metadata_row.id,
+            ).delete()
```

Comment:
As mentioneded on IRC this function should do the cleanup completely on the db side to prevent having to fetch the data to delete it afterwards. In my experience such cleanup function tend to break with race conditions if the list of things to delete exceeds memory or runtime limits.